## 1. Reference file

Reference file is needed in the **_scan_** module, and the version and contig name of reference you used should be
in accordance with your bam files. And the reference should be save as uncompressed fasta (fa) format.

## 2. microsatellite file

This file is a tab-separated table and generated by scan or baseline command. Users can delete or merge this file on
their own,
as long as the first row remains unchanged and the other rows follow the specified rules.
Only the rows where the last column contains a PASS value will be used for further analysis.

* chromosome: chromosome id
* location: start position of the microsatellite
* repeat_unit_length: repeat unit length
* repeat_times: repeat times
* repeat_unit_bases: content of the repeat unit
* left_flank_bases: prefix sequences of the microsatellite
* right_flank_bases: suffix sequences of the microsatellite
* threshold: baseline of error (deletion,_p_), -1 indicates that this value is unavailable.
* support_num: sample number for baseline building of the site, -1 indicates that this value is unavailable.
* filter: Only the sites with a PASS value will proceed with the subsequent operations.

**Example:**

| chromosome | location | repeat_unit_length | repeat_times | repeat_unit_bases | left_flank_bases | right_flank_bases | threshold | support_num | filter |
|------------|----------|--------------------|--------------|-------------------|------------------|-------------------|-----------|-------------|--------|
| chr1       | 3780974  | 1                  | 11           | A                 | ATCTC            | CCAAC             | 0.080981  | 30          | PASS   |
| chr1       | 3784993  | 1                  | 13           | A                 | TCTCC            | GTTCG             | 0.007576  | 18          | PASS   |
| chr1       | 3836468  | 1                  | 24           | T                 | CCCCG            | ACCAA             | 0.061750  | 19          | PASS   |
| chr1       | 3872414  | 1                  | 13           | A                 | TCAAG            | GAGAC             | 0.02842   | 28          | PASS   |
| chr1       | 4712522  | 3                  | 7            | CCA               | GGCCG            | CGGCC             | 0.024391  | 3           | PASS   |

## 3. distribution file (*_dis)

Every two or three lines represent information about a site. The first line provides the basic information about the
site, including the chromosome, position, prefix sequences, the repeat pattern on the reference genome,
and suffix sequences.

The second and third lines represent the distribution of repeat lengths at the site.
Specifically, the part before the colon represents the sample type(T: tumor N: normal),
and the number after the colon indicates the number of reads supporting the repeat at that position.

For example, the value of 2 in the second column indicates that in the normal sample,
there are 2 reads supporting 4 (the forth position) repeats at position chr1_1090291,
and 72 means there are 72 reads supporting 5 (the five position) repeats at that site in the sample.

**Example:**

    chr1 1090291 CTGGG 5[CA] CCAGT
    N: 0 0 0 2 72 0 0 0 0 0 ...
    T: 0 0 0 11 26 0 0 0 0 0 ...
    chr1 1043873 GGGCA 5[GT] CCCCG
    N: 0 0 0 1 68 0 0 0 0 0 0 0 0 0
    T: 0 0 0 0 54 0 0 0 0 0 0 0 0 0

## 4. output microsatellite information of msi command (*_all / \*_unstable)

difference:  the chi-squared test statistic.
P_value: the chi-squared _P_ value.
FDR:  Corrected FDR (False Discovery Rate).

| chromosome | location | left_flank_bases | repeat_times | repeat_unit_bases | right_flank_bases | difference | P_value   | FDR     |
|------------|----------|------------------|--------------|-------------------|-------------------|------------|-----------|---------|
| chr1       | 1090291  | CTGGG            | 5            | CA                | CCAGT             | 0.42553    | 0.0052775 | 0.06333 |
| chr1       | 1398450  | AGGGG            | 9            | A                 | GTCAA             | 0.16       | 0.19533   | 1.172   |
| chr1       | 1180841  | GTGCC            | 8            | G                 | TCCAG             | 0.11765    | 0.30163   | 1.2065  |
| chr1       | 1420960  | CCTGT            | 5            | TCC               | TCTCT             | 0.043497   | 0.34924   | 1.0477  |

## 5. output microsatellite information of pro command (*_all / \*_unstable）

* pro_p : probability of deletion of this site.
* pro_q : probability of insertion of this site.     
* CovReads: read number covered this site. 
* threshold: The baseline of this site; sites with pro_p greater than this value will be classified as unstable loci.

**Example** 

| chromosome | location | left_flank_bases | repeat_times | repeat_unit_bases | right_flank_bases | pro_p    | pro_q    | CovReads | threshold |
|------------|----------|------------------|--------------|-------------------|-------------------|----------|----------|----------|-----------|
| chr1       | 1043873  | GGGCA            | 5            | GT                | CCCCG             | 0.000000 | 0.000000 | 54       | 0.000000  |
| chr1       | 1046479  | GCCGG            | 8            | C                 | GGCGC             | 0.000000 | 0.003195 | 39       | 0.012030  |
| chr1       | 1297645  | GGCAC            | 5            | GT                | GCACA             | 0.004878 | 0.000000 | 41       | 0.000000  |
| chr1       | 1395518  | CTGCC            | 5            | AG                | CACAG             | 0.000000 | 0.000000 | 20       | 0.002737  |
| chr1       | 1398450  | AGGGG            | 9            | A                 | GTCAA             | 0.000000 | 0.000000 | 18       | 0.000000  |

## 6. output microsatellite information of pro command (*_all / \*_unstable）

| Total_Number_of_Sites | Number_of_Unstable_Sites | %    |
|-----------------------|--------------------------|------|
| 12                    | 0                        | 0.00 |

## 7. configure file of baseline module**

Configure file is needed in baseline module, the first column is the sample name and the second is the absolute path of
its pro output of normal sample.

**Example:**

    case1   /path/to/normal/case1_all
    case2   /path/to/normal/case2_all
    case3   /path/to/normal/case3_all


